### DISCUSSION TITLE
Almost there: Getting clangd remote indexing to work across multiple checkouts

### DISCUSSION THREAD

**Discussion by breadsander** [clangd/clangd#2065] [Q&A] [answered]
I've spent the last week or some trying to wrangle all the necessary bits and pieces to run a clangd-index-server and use it across multiple checkouts. I have feeling I am nearly there, but I can't get across this final hump.

### Context
The codebase that I'm generating the index has millions of lines of C/C++. It is a large enterprise monorepro. I want to setup a nightly run that takes the current head-of-line, compiles the clang index, and hosts an index server that can be attached to. I want to be able to checkout multiple instances of the codebase (even ones that have moved forward from the nightly) and have them connect to the remote-index as a "baseline". I am using Emacs + Eglot to handle the clangd client->server connection.

The instructions and documentation I've found online suggests this should be possible.

### What I Have Working
I am able to generate a compile_commands.json since we utilize clang for our compilation. It is able to spit out the commands and I can collect them into a large compile_commands.json file.

I am able to compile an index via "clangd-indexer --executor=all-TUs compile_commands.json > clangd.dex" and it works successfully.

I am able to start the clangd-index-server succesfully.

**I am able to set up a clangd.yaml for exactly-one checkout to connect to the remote-index**. It is set in "~/.config/clangd/config.yaml". The YAML looks something like this:
```
If:
  PathMatch: /an/absolute/path/to/client/checkoutA/.*
Index:
  External:
    Server: my.server.host:50051
    MountPoint: /an/absolute/path/to/server/checkoutA/
```
Important note: As of right now "PathMatch" and "MountPoint" are technically the same exact path, since the path is a common NFS mount between the client and server. The client and server **are on different machines** The future plan is for the "MountPoint" to be "/an/absolute/path/to/nightly" at some point in the future.

Lastly, I am able to connect to the clangd remote index and things work flawlessly. No background indexing. And, at the server I can see the LSP requests land. So far so good-- checkoutA works.

### What Isn't Working
Next, I want to get a second checkout-- this time let's call it checkoutB. I edit my config.yaml to this:
```
If:
  PathMatch: /an/absolute/path/to/client/checkoutA/.*
Index:
  External:
    Server: my.server.host:50051
    MountPoint: /an/absolute/path/to/server/checkoutA/
---
If:
  PathMatch: /an/absolute/path/to/client/checkoutB/.*
Index:
  External:
    Server: my.server.host:50051
    MountPoint: /an/absolute/path/to/server/checkoutA/
```
Notice I basically say that checkoutB should look at checkoutA as well-- even if checkoutB is further ahead of checkoutA. Note now that the PathMatch and MountPoint paths are distinct from one another.

In the background I have taken the compile_flags.json from client checkoutA and copied it to client checkoutB, and I did a string replace of the entries to change the absolute paths of checkoutA to the absolute paths of checkoutB.

However, when I start eglot, I do not connect to the clangd remote index and instead start a background index of clientB. The clientB compile_commands.json appear to work as the local-clangd does on-the-fly compilation and parsing just fine. But, I cannot figure out why the checkoutB refuses to just go to the remote index.

### Theories
My mind keeps gravitating towards the use of absolute paths in both the yaml and the compile_commands.json. I can't help but wonder if by using absolute paths some translation isn't happening and the clangd-remote-index sees the connection as not compatible. Or maybe the clangd-remote-index isn't even contacted and something about is causing the local clangd not to attempt to use the external index.

In particular, I am confused about the MountPoint. Here are a few snippets about remote-index MountPoint that I've found in the documentation.
> MountPoint can be used to specify source root for the index. This is necessary to handle relative path conversions. 

and also
>MountPoint defaults to location of the config fragment if not provided, must be absolute in global config and relative in local config.

with the global config being
> user configuration: a config.yaml file in an OS-specific directory:
Windows: %LocalAppData%\clangd\config.yaml, typically C:\Users\Bob\AppData\Local\clangd\config.yaml.
macOS: ~/Library/Preferences/clangd/config.yaml
Linux and others: $XDG_CONFIG_HOME/clangd/config.yaml, typically ~/.config/clangd/config.yaml.

So... configs for a remote-index must be set in the user (global) config. And global configs must use an absolute path for MountPoints.  Therefore, all remote-index configs must use an absolute MountPath. But... they have something to do with relative-path conversions? This has me very turned around.

Anyway, if anyone can help me figure out the final piece here I'd be very very grateful. I really think I am so close at this point.

Edit: I should add that I am starting the clangd-index-server via:
```
clangd-index-server /an/absolute/path/to/server/checkoutA/clang.dex /an/absolute/path/to/server/checkoutA
```

**Comment by breadsander** [comment, marked as answer]
Well, isn't there a saying about how when you finally ask a question the answer comes to you? The problem was precisely the mountpoint config:
```
If:
  PathMatch: /an/absolute/path/to/client/checkoutA/.*
Index:
  External:
    Server: my.server.host:50051
    MountPoint: /an/absolute/path/to/server/checkoutA/
---
If:
  PathMatch: /an/absolute/path/to/client/checkoutB/.*
Index:
  External:
    Server: my.server.host:50051
    MountPoint: /an/absolute/path/to/server/checkoutA/
```
For some reason I had thought MountPoint was supposed to be a path at the server. It's not. MountPoint is the clue given to the **client** clangd on how to treat the index handed back to it by the remote-index server.

I think in most simple cases the setup will be:
```
If:
  PathMatch: /my/path/to/checkoutA/.*
Index:
  External:
    Server: my.server.host:50051
    MountPoint: /my/path/to/checkouA/
---
If:
  PathMatch: /my/path/to/checkoutB/.*
Index:
  External:
    Server: my.server.host:50051
    MountPoint: /my/path/to/checkoutB/
---
<repeat for further checkouts>
```
Everything is working now. This has been a white whale for a long time for me and I am absolutely thrilled with how well this works.

I am going to leave this dicussion opened to see if maybe a documentation adjustment is needed that explains more clearly what MountPoint is. Thanks.

**Reply by HighCommander4** [reply]
Glad you got it to work!

> I am going to leave this dicussion opened to see if maybe a documentation adjustment is needed that explains more clearly what MountPoint is.

Yeah, it would probably help to add a clarifying comment at https://clangd.llvm.org/guides/remote-index#configuring-clangd-to-use-the-remote-index or https://clangd.llvm.org/config#external (or both).

Perhaps you might be interested in sending a documentation patch? The repo for the docs is https://github.com/llvm/clangd-www/.
